File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

    int maxDiam = 0;

    int diameterOfBinaryTree(TreeNode* root) {

        maxHeight(root);
        return maxDiam;
        
       

    }

    int maxHeight(TreeNode* node) {
        if(node == nullptr) {
            return 0;
        }

        int depth = 1;

        int left = maxHeight(node->left);
        int right = maxHeight(node->right);

        depth = depth + max(left, right);

        maxDiam = max(maxDiam, left + right);

        return depth;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        res = 0

        def dfs(root, greatest):
            nonlocal res

            if not root:
                return

            if root.val >= greatest:
                res+=1
            
            greatest = max(root.val, greatest)

            dfs(root.left, greatest)
            dfs(root.right, greatest)

        
        dfs(root, -101)

        return res